### PR TITLE
authenticate/providers : add gitlab support

### PR DESCRIPTION
**Pull request by desimone**
**TL;DR** This adds Gitlab (onprem/hosted) support to pomerium. But if you use a validator that depends on validating an email (the current default), it probably won't work unless everyone in your org has made their email public, or you write your own custom validator. 

---

This was a pretty frustrating one! Much of the documentation was outdated, or wrong. The silver lining is that it did get me thinking more about how to improve authentication/authorization away from email centric, as well as to incorporate additional user information (via `userinfo_endpoint`). 

Unlike other providers, gitlab **does not supply a user's email** during OpenID Connect. Even a secondary call to the user information endpoint does not contain the user's email unless the user explicitly sets her email as public. 

As pomerium is currently very email centric, I would caution against using Gitlab until they fix their OpenID Connect implementation as discussed [here](https://gitlab.com/gitlab-org/gitlab-ce/issues/44435#note_88150387).

I've added docs for [Gitlab](https://www.pomerium.io/guide/identity-providers.html#gitlab).
I've also added a docker-compse example script that hopefully will help save you a few hours trying to bring up an on-premise version. 

```docker
version: "3"

services:
  # NGINX routes to pomerium's services depending on the request.
  nginx:
    image: jwilder/nginx-proxy:latest
    ports:
      - "443:443"
    volumes:
      # NOTE!!! : nginx must be supplied with your wildcard certificates. And it expects
      # it in the format of whatever your wildcard domain name is in.
      # see : https://github.com/jwilder/nginx-proxy#wildcard-certificates
      # So, if your subdomain is corp.beyondperimeter.com, you'd have the following :
      - ./cert.pem:/etc/nginx/certs/corp.beyondperimeter.com.crt:ro
      - ./privkey.pem:/etc/nginx/certs/corp.beyondperimeter.com.key:ro
      - /var/run/docker.sock:/tmp/docker.sock:ro

  pomerium-authenticate:
    build: .
    restart: always
    depends_on:
      - "gitlab"
    environment:
      - POMERIUM_DEBUG=true
      - SERVICES=authenticate
      # auth settings
      - REDIRECT_URL=https://sso-auth.corp.beyondperimeter.com/oauth2/callback
      - IDP_PROVIDER="gitlab"
      - IDP_PROVIDER_URL=https://gitlab.corp.beyondperimeter.com
      - IDP_CLIENT_ID=022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76
      - IDP_CLIENT_SECRET=fb7598c520c346915ee369eee57688938fe4f31329a308c4669074da562714b2
      - PROXY_ROOT_DOMAIN=beyondperimeter.com
      - ALLOWED_DOMAINS=*
      - SKIP_PROVIDER_BUTTON=false
      # shared service settings
      # Generate 256 bit random keys  e.g. `head -c32 /dev/urandom | base64`
      - SHARED_SECRET=aDducXQzK2tPY3R4TmdqTGhaYS80eGYxcTUvWWJDb2M=
      - COOKIE_SECRET=V2JBZk0zWGtsL29UcFUvWjVDWWQ2UHExNXJ0b2VhcDI=
      - VIRTUAL_PROTO=https
      - VIRTUAL_HOST=sso-auth.corp.beyondperimeter.com
      - VIRTUAL_PORT=443
    volumes: # volumes is optional; used if passing certificates as files
      - ./cert.pem:/pomerium/cert.pem:ro
      - ./privkey.pem:/pomerium/privkey.pem:ro
    expose:
      - 443

  pomerium-proxy:
    build: .
    restart: always
    environment:
      - POMERIUM_DEBUG=true
      - SERVICES=proxy
      # proxy settings
      - AUTHENTICATE_SERVICE_URL=https://sso-auth.corp.beyondperimeter.com
      - ROUTES=https://httpbin.corp.beyondperimeter.com=http://httpbin,https://hello.corp.beyondperimeter.com=http://hello-world/
      # Generate 256 bit random keys  e.g. `head -c32 /dev/urandom | base64`
      - SHARED_SECRET=aDducXQzK2tPY3R4TmdqTGhaYS80eGYxcTUvWWJDb2M=
      - COOKIE_SECRET=V2JBZk0zWGtsL29UcFUvWjVDWWQ2UHExNXJ0b2VhcDI=
      - SIGNING_KEY=LS0tLS1CRUdJTiBFQyBQUklWQVRFIEtFWS0tLS0tCk1IY0NBUUVFSU0zbXBaSVdYQ1g5eUVneFU2czU3Q2J0YlVOREJTQ0VBdFFGNWZVV0hwY1FvQW9HQ0NxR1NNNDkKQXdFSG9VUURRZ0FFaFBRditMQUNQVk5tQlRLMHhTVHpicEVQa1JyazFlVXQxQk9hMzJTRWZVUHpOaTRJV2VaLwpLS0lUdDJxMUlxcFYyS01TYlZEeXI5aWp2L1hoOThpeUV3PT0KLS0tLS1FTkQgRUMgUFJJVkFURSBLRVktLS0tLQo=
      # nginx settings
      - VIRTUAL_PROTO=https
      - VIRTUAL_HOST=*.corp.beyondperimeter.com
      - VIRTUAL_PORT=443
    volumes: # volumes is optional; used if passing certificates as files
      - ./cert.pem:/pomerium/cert.pem:ro
      - ./privkey.pem:/pomerium/privkey.pem:ro
    expose:
      - 443

  # https://httpbin.corp.beyondperimeter.com
  httpbin:
    image: kennethreitz/httpbin:latest
    expose:
      - 80
  # https://hello.corp.beyondperimeter.com
  hello-world:
    image: tutum/hello-world:latest
    expose:
      - 80
  gitlab:
    hostname: gitlab.corp.beyondperimeter.com
    image: gitlab/gitlab-ce:latest
    restart: always
    expose:
      - 443
      - 80
      - 22
    environment:
      GITLAB_OMNIBUS_CONFIG: |
        external_url 'https://gitlab.corp.beyondperimeter.com'
        nginx['ssl_certificate'] = '/etc/gitlab/trusted-certs/corp.beyondperimeter.com.crt'
        nginx['ssl_certificate_key'] = '/etc/gitlab/trusted-certs/corp.beyondperimeter.com.key'
      # nginx settings
      VIRTUAL_PROTO: https
      VIRTUAL_HOST: gitlab.corp.beyondperimeter.com
      VIRTUAL_PORT: 443
    volumes:
      - ./cert.pem:/etc/gitlab/trusted-certs/corp.beyondperimeter.com.crt
      - ./privkey.pem:/etc/gitlab/trusted-certs/corp.beyondperimeter.com.key
      - $HOME/gitlab/config:/etc/gitlab
      - $HOME/gitlab/logs:/var/log/gitlab
      - $HOME/gitlab/data:/var/opt/gitlab

``` 

I think @mbarrien (#26) and I did a little duplicated work trying to implement #20 .

See: 
- 
- https://issues.jenkins-ci.org/browse/JENKINS-44121
- https://docs.gitlab.com/ee/integration/openid_connect_provider.html
- https://docs.gitlab.com/ee/integration/oauth_provider.html
- https://gitlab.com/.well-known/openid-configuration

### CODE REVIEW

**Comment by desimone** [comment]
Ugh. Spent most of today trying to solve this. I think the issue is Gitlab. It took me several hours to get gitlab configured correctly locally to try this out. See this docker-compose if you want to [try it out](https://github.com/pomerium/pomerium/blob/c8287c117c7f2108f2e898324e4b00ce9513537d/docs/examples/gitlab.docker-compose.yml).

Maybe @mbarrien @fnkr have some ideas?

/var/log/gitlab/gitlab-rails/production.log
```
gitlab_1                 | 172.18.0.3 - - [24/Jan/2019:00:13:44 +0000] "GET /oauth/undefined HTTP/1.1" 404 6475 "https://gitlab.corp.beyondperimeter.com/oauth/authorize" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production_json.log <==
gitlab_1                 | {"method":"GET","path":"/oauth/authorize","format":"html","controller":"Oauth::AuthorizationsController","action":"new","status":200,"duration":46.9,"view":28.05,"db":11.94,"time":"2019-01-24T00:13:44.589Z","params":[{"key":"access_type","value":"offline"},{"key":"approval_prompt","value":"force"},{"key":"client_id","value":"022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76"},{"key":"redirect_uri","value":"https://sso-auth.corp.beyondperimeter.com/oauth2/callback"},{"key":"response_type","value":"code"},{"key":"scope","value":"openid profile email"},{"key":"state","value":"NmQ2ZjVlNGU2M2JhMWQyZDYzZTQ2NjU4ZjE2M2ZlMjlkZTdmNWFmY2Y5NTA5MjVjNmUzMGI4YTljNTYyYzVhYzpodHRwczovL3Nzby1hdXRoLmNvcnAuYmV5b25kcGVyaW1ldGVyLmNvbS9zaWduX2luP3JlZGlyZWN0X3VyaT1odHRwcyUzQSUyRiUyRmh0dHBiaW4uY29ycC5iZXlvbmRwZXJpbWV0ZXIuY29tJTJGLnBvbWVyaXVtJTJGY2FsbGJhY2smcmVzcG9uc2VfdHlwZT1jb2RlJnNoYXJlZF9zZWNyZXQ9YURkdWNYUXpLMnRQWTNSNFRtZHFUR2hhWVM4MGVHWXhjVFV2V1dKRGIyTSUzRCZzaWc9MU9Gby02NGRLN0xUQTFMQzUzRzJZQUNNUWdNVXRWb1NWQ0F3UzQ1LTBZZyUzRCZzdGF0ZT04X3BTbU5RbGkyTm1zVmJaRDBIdXp1RXBESGExSEkxNUxBOE9JN1M3dGpBQm5yMElxZ0xzVW5LY3JoTTVNeU1ZcTd4cDR0OTZxTGRWdkpFQUJKMFR5cS1PVDUyMGFWRWlabXVpNVBsZkNqV2hBZjZSVmU0R2RzQUpnRjI1cERxZGhidG5aeEszd1l4MC1iRUtVbTdxeGkzR2ZGUnpjTDY0RGc3eWszVlhwVUVBcF90UXdXeVNEVktZLXZVRE85YU1LUU1TZjBPMzVRJnRzPTE1NDgyODg4MjI="}],"remote_ip":null,"user_id":null,"username":null,"ua":null,"correlation_id":"IG9sKjmLJk6"}
gitlab_1                 | {"method":"GET","path":"/oauth/undefined","format":"html","controller":"ApplicationController","action":"route_not_found","status":404,"duration":16.57,"view":1.09,"db":7.35,"time":"2019-01-24T00:13:44.917Z","params":[{"key":"unmatched_route","value":"oauth/undefined"}],"remote_ip":"172.18.0.3","user_id":1,"username":"root","ua":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36","correlation_id":"p1XxtdJRNH9"}
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production.log <==
gitlab_1                 | Started GET "/oauth/authorize?access_type=offline&approval_prompt=force&client_id=022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76&redirect_uri=https%3A%2F%2Fsso-auth.corp.beyondperimeter.com%2Foauth2%2Fcallback&response_type=code&scope=openid+profile+email&state=NmQ2ZjVlNGU2M2JhMWQyZDYzZTQ2NjU4ZjE2M2ZlMjlkZTdmNWFmY2Y5NTA5MjVjNmUzMGI4YTljNTYyYzVhYzpodHRwczovL3Nzby1hdXRoLmNvcnAuYmV5b25kcGVyaW1ldGVyLmNvbS9zaWduX2luP3JlZGlyZWN0X3VyaT1odHRwcyUzQSUyRiUyRmh0dHBiaW4uY29ycC5iZXlvbmRwZXJpbWV0ZXIuY29tJTJGLnBvbWVyaXVtJTJGY2FsbGJhY2smcmVzcG9uc2VfdHlwZT1jb2RlJnNoYXJlZF9zZWNyZXQ9YURkdWNYUXpLMnRQWTNSNFRtZHFUR2hhWVM4MGVHWXhjVFV2V1dKRGIyTSUzRCZzaWc9MU9Gby02NGRLN0xUQTFMQzUzRzJZQUNNUWdNVXRWb1NWQ0F3UzQ1LTBZZyUzRCZzdGF0ZT04X3BTbU5RbGkyTm1zVmJaRDBIdXp1RXBESGExSEkxNUxBOE9JN1M3dGpBQm5yMElxZ0xzVW5LY3JoTTVNeU1ZcTd4cDR0OTZxTGRWdkpFQUJKMFR5cS1PVDUyMGFWRWlabXVpNVBsZkNqV2hBZjZSVmU0R2RzQUpnRjI1cERxZGhidG5aeEszd1l4MC1iRUtVbTdxeGkzR2ZGUnpjTDY0RGc3eWszVlhwVUVBcF90UXdXeVNEVktZLXZVRE85YU1LUU1TZjBPMzVRJnRzPTE1NDgyODg4MjI%3D" for 127.0.0.1 at 2019-01-24 00:13:44 +0000
gitlab_1                 | Processing by Oauth::AuthorizationsController#new as HTML
gitlab_1                 |   Parameters: {"access_type"=>"offline", "approval_prompt"=>"force", "client_id"=>"022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76", "redirect_uri"=>"https://sso-auth.corp.beyondperimeter.com/oauth2/callback", "response_type"=>"code", "scope"=>"openid profile email", "state"=>"NmQ2ZjVlNGU2M2JhMWQyZDYzZTQ2NjU4ZjE2M2ZlMjlkZTdmNWFmY2Y5NTA5MjVjNmUzMGI4YTljNTYyYzVhYzpodHRwczovL3Nzby1hdXRoLmNvcnAuYmV5b25kcGVyaW1ldGVyLmNvbS9zaWduX2luP3JlZGlyZWN0X3VyaT1odHRwcyUzQSUyRiUyRmh0dHBiaW4uY29ycC5iZXlvbmRwZXJpbWV0ZXIuY29tJTJGLnBvbWVyaXVtJTJGY2FsbGJhY2smcmVzcG9uc2VfdHlwZT1jb2RlJnNoYXJlZF9zZWNyZXQ9YURkdWNYUXpLMnRQWTNSNFRtZHFUR2hhWVM4MGVHWXhjVFV2V1dKRGIyTSUzRCZzaWc9MU9Gby02NGRLN0xUQTFMQzUzRzJZQUNNUWdNVXRWb1NWQ0F3UzQ1LTBZZyUzRCZzdGF0ZT04X3BTbU5RbGkyTm1zVmJaRDBIdXp1RXBESGExSEkxNUxBOE9JN1M3dGpBQm5yMElxZ0xzVW5LY3JoTTVNeU1ZcTd4cDR0OTZxTGRWdkpFQUJKMFR5cS1PVDUyMGFWRWlabXVpNVBsZkNqV2hBZjZSVmU0R2RzQUpnRjI1cERxZGhidG5aeEszd1l4MC1iRUtVbTdxeGkzR2ZGUnpjTDY0RGc3eWszVlhwVUVBcF90UXdXeVNEVktZLXZVRE85YU1LUU1TZjBPMzVRJnRzPTE1NDgyODg4MjI="}
gitlab_1                 | Completed 200 OK in 46ms (Views: 28.1ms | ActiveRecord: 11.9ms)
gitlab_1                 | Started GET "/oauth/undefined" for 127.0.0.1 at 2019-01-24 00:13:44 +0000
gitlab_1                 | Processing by ApplicationController#route_not_found as HTML
gitlab_1                 |   Parameters: {"unmatched_route"=>"oauth/undefined"}
gitlab_1                 | Completed 404 Not Found in 15ms (Views: 1.1ms | ActiveRecord: 7.3ms)
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production_json.log <==
gitlab_1                 | {"method":"GET","path":"/-/metrics","format":"html","controller":"MetricsController","action":"index","status":200,"duration":3.73,"view":0.29,"db":0.0,"time":"2019-01-24T00:13:56.695Z","params":[],"remote_ip":null,"user_id":null,"username":null,"ua":null,"correlation_id":"3dca4419-c0ef-424a-b2e8-7d885ab875c7"}
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production.log <==
gitlab_1                 | Started GET "/-/metrics" for 127.0.0.1 at 2019-01-24 00:13:56 +0000
gitlab_1                 | Processing by MetricsController#index as HTML
gitlab_1                 | Completed 200 OK in 2ms (Views: 0.3ms | ActiveRecord: 0.0ms)
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/sidekiq_exporter.log <==
gitlab_1                 | [2019-01-24 00:13:56] 127.0.0.1 - - [24/Jan/2019:00:13:56 UTC] "GET /metrics HTTP/1.1" 200 1224 "-" "Prometheus/2.5.0"
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production.log <==
gitlab_1                 | Started GET "/help" for 127.0.0.1 at 2019-01-24 00:14:06 +0000
gitlab_1                 | Processing by HelpController#index as */*
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-workhorse/current <==
gitlab_1                 | 2019-01-24_00:14:06.53595 localhost 127.0.0.1:0 - - [2019/01/24:00:14:06 +0000] "GET /help HTTP/1.1" 200 62045 "" "curl/7.59.0-DEV" 0.357
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/nginx/gitlab_access.log <==
gitlab_1                 | 127.0.0.1 - - [24/Jan/2019:00:14:06 +0000] "GET /help HTTP/1.1" 200 62098 "" "curl/7.59.0-DEV"
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production_json.log <==
gitlab_1                 | {"method":"GET","path":"/help","format":"*/*","controller":"HelpController","action":"index","status":200,"duration":312.61,"view":303.51,"db":2.05,"time":"2019-01-24T00:14:06.219Z","params":[],"remote_ip":"127.0.0.1","user_id":null,"username":null,"ua":"curl/7.59.0-DEV","correlation_id":"olRpSG6in2"}
gitlab_1                 |
gitlab_1                 | ==> /var/log/gitlab/gitlab-rails/production.log <==
```

/var/log/gitlab/gitlab-rails/production.log
```
Processing by Admin::ApplicationsController#update as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"[FILTERED]", "doorkeeper_application"=>{"name"=>"pomerium", "redirect_uri"=>"https://sso-auth.corp.beyondperimeter.com/oauth2/callback", "trusted"=>"1", "scopes"=>["api", "read_user", "sudo", "read_repository", "openid"]}, "id"=>"1"}
Redirected to https://gitlab.corp.beyondperimeter.com/admin/applications/1
Completed 302 Found in 45ms (ActiveRecord: 17.9ms)
Started GET "/admin/applications/1" for 127.0.0.1 at 2019-01-24 00:13:38 +0000
Processing by Admin::ApplicationsController#show as HTML
  Parameters: {"id"=>"1"}
Completed 200 OK in 47ms (Views: 27.4ms | ActiveRecord: 12.4ms)
Started GET "/-/metrics" for 127.0.0.1 at 2019-01-24 00:13:41 +0000
Processing by MetricsController#index as HTML
Completed 200 OK in 3ms (Views: 0.4ms | ActiveRecord: 0.0ms)
Started GET "/oauth/authorize?access_type=offline&approval_prompt=force&client_id=022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76&redirect_uri=https%3A%2F%2Fsso-auth.corp.beyondperimeter.com%2Foauth2%2Fcallback&response_type=code&scope=openid+profile+email&state=NmQ2ZjVlNGU2M2JhMWQyZDYzZTQ2NjU4ZjE2M2ZlMjlkZTdmNWFmY2Y5NTA5MjVjNmUzMGI4YTljNTYyYzVhYzpodHRwczovL3Nzby1hdXRoLmNvcnAuYmV5b25kcGVyaW1ldGVyLmNvbS9zaWduX2luP3JlZGlyZWN0X3VyaT1odHRwcyUzQSUyRiUyRmh0dHBiaW4uY29ycC5iZXlvbmRwZXJpbWV0ZXIuY29tJTJGLnBvbWVyaXVtJTJGY2FsbGJhY2smcmVzcG9uc2VfdHlwZT1jb2RlJnNoYXJlZF9zZWNyZXQ9YURkdWNYUXpLMnRQWTNSNFRtZHFUR2hhWVM4MGVHWXhjVFV2V1dKRGIyTSUzRCZzaWc9MU9Gby02NGRLN0xUQTFMQzUzRzJZQUNNUWdNVXRWb1NWQ0F3UzQ1LTBZZyUzRCZzdGF0ZT04X3BTbU5RbGkyTm1zVmJaRDBIdXp1RXBESGExSEkxNUxBOE9JN1M3dGpBQm5yMElxZ0xzVW5LY3JoTTVNeU1ZcTd4cDR0OTZxTGRWdkpFQUJKMFR5cS1PVDUyMGFWRWlabXVpNVBsZkNqV2hBZjZSVmU0R2RzQUpnRjI1cERxZGhidG5aeEszd1l4MC1iRUtVbTdxeGkzR2ZGUnpjTDY0RGc3eWszVlhwVUVBcF90UXdXeVNEVktZLXZVRE85YU1LUU1TZjBPMzVRJnRzPTE1NDgyODg4MjI%3D" for 127.0.0.1 at 2019-01-24 00:13:44 +0000
Processing by Oauth::AuthorizationsController#new as HTML
  Parameters: {"access_type"=>"offline", "approval_prompt"=>"force", "client_id"=>"022dbbd09402441dc7af1924b679bc5e6f5bf0d7a555e55b38c51e2e4e6cee76", "redirect_uri"=>"https://sso-auth.corp.beyondperimeter.com/oauth2/callback", "response_type"=>"code", "scope"=>"openid profile email", "state"=>"NmQ2ZjVlNGU2M2JhMWQyZDYzZTQ2NjU4ZjE2M2ZlMjlkZTdmNWFmY2Y5NTA5MjVjNmUzMGI4YTljNTYyYzVhYzpodHRwczovL3Nzby1hdXRoLmNvcnAuYmV5b25kcGVyaW1ldGVyLmNvbS9zaWduX2luP3JlZGlyZWN0X3VyaT1odHRwcyUzQSUyRiUyRmh0dHBiaW4uY29ycC5iZXlvbmRwZXJpbWV0ZXIuY29tJTJGLnBvbWVyaXVtJTJGY2FsbGJhY2smcmVzcG9uc2VfdHlwZT1jb2RlJnNoYXJlZF9zZWNyZXQ9YURkdWNYUXpLMnRQWTNSNFRtZHFUR2hhWVM4MGVHWXhjVFV2V1dKRGIyTSUzRCZzaWc9MU9Gby02NGRLN0xUQTFMQzUzRzJZQUNNUWdNVXRWb1NWQ0F3UzQ1LTBZZyUzRCZzdGF0ZT04X3BTbU5RbGkyTm1zVmJaRDBIdXp1RXBESGExSEkxNUxBOE9JN1M3dGpBQm5yMElxZ0xzVW5LY3JoTTVNeU1ZcTd4cDR0OTZxTGRWdkpFQUJKMFR5cS1PVDUyMGFWRWlabXVpNVBsZkNqV2hBZjZSVmU0R2RzQUpnRjI1cERxZGhidG5aeEszd1l4MC1iRUtVbTdxeGkzR2ZGUnpjTDY0RGc3eWszVlhwVUVBcF90UXdXeVNEVktZLXZVRE85YU1LUU1TZjBPMzVRJnRzPTE1NDgyODg4MjI="}
Completed 200 OK in 46ms (Views: 28.1ms | ActiveRecord: 11.9ms)
Started GET "/oauth/undefined" for 127.0.0.1 at 2019-01-24 00:13:44 +0000
Processing by ApplicationController#route_not_found as HTML
  Parameters: {"unmatched_route"=>"oauth/undefined"}
Completed 404 Not Found in 15ms (Views: 1.1ms | ActiveRecord: 7.3ms)
Started GET "/-/metrics" for 127.0.0.1 at 2019-01-24 00:13:56 +0000
Processing by MetricsController#index as HTML
Completed 200 OK in 2ms (Views: 0.3ms | ActiveRecord: 0.0ms)
Started GET "/help" for 127.0.0.1 at 2019-01-24 00:14:06 +0000
Processing by HelpController#index as */*
Completed 200 OK in 310ms (Views: 303.5ms | ActiveRecord: 2.1ms)
Started GET "/-/metrics" for 127.0.0.1 at 2019-01-24 00:14:11 +0000
Processing by MetricsController#index as HTML
Completed 200 OK in 2ms (Views: 0.4ms | ActiveRecord: 0.0ms)
Started GET "/-/metrics" for 127.0.0.1 at 2019-01-24 00:14:26 +0000
Processing by MetricsController#index as HTML
Completed 200 OK in 3ms (Views: 0.5ms | ActiveRecord: 0.0ms)
```
